app/forms/apply_trainee_data_form.rb

```diff
@@ -16,24 +16,52 @@ def validator(name, options)
   validator :diversity, form: "Diversities::FormValidator"
   validator :degrees, form: "DegreesForm"
 
+  delegate :apply_application?, to: :trainee
+
   attr_accessor :mark_as_reviewed
 
-  def initialize(trainee:)
+  def initialize(trainee)
     @trainee = trainee
   end
 
   def save
+    return unless all_forms_valid?
+
     trainee.progress.personal_details = true
     trainee.progress.contact_details = true
     trainee.progress.diversity = true
     trainee.progress.degrees = true
     trainee.save!
   end
 
+  def all_forms_valid?
+    form_validators.all? do |section|
+      form(section[1]).valid?
+    end
+  end
+
+  def form(section)
+    section[:form].constantize.new(trainee)
+  end
+
   def progress_status(progress_key)
     progress_service(progress_key).status.parameterize(separator: "_").to_sym
   end
 
+  def progress
+    form_validators.all? do |section_key|
+      trainee.progress.public_send(section_key[0])
+    end
+  end
+
+  def fields
```

Comment:
private? or test? or in use?

Reply:
In use by ProgressService

https://github.com/DFE-Digital/register-trainee-teachers/blob/05275fa3b4070e34986e9418f420e94a777364d7/app/services/progress_service.rb#L23-L25

Reply:
:+1: